lib/Models/ThreddsCatalogGroup.ts

```diff
@@ -0,0 +1,205 @@
+import GroupMixin from "../ModelMixins/GroupMixin";
+import UrlMixin from "../ModelMixins/UrlMixin";
+import ThreddsCatalogGroupTraits from "../Traits/ThreddsCatalogGroupTraits";
+import ModelReference from "../Traits/ModelReference";
+import CatalogGroup from "./CatalogGroupNew";
+import CreateModel from "./CreateModel";
+import LoadableStratum from "./LoadableStratum";
+import { BaseModel } from "./Model";
+import StratumOrder from "./StratumOrder";
+import ThreddsItemReference from "./ThreddsItemReference";
+import WebMapServiceCatalogGroup from "./WebMapServiceCatalogGroup";
+import CommonStrata from "./CommonStrata";
+
+interface ThreddsCatalog {
+  id: string;
+  name: string;
+  title: string;
+  isLoaded: boolean;
+  url: string;
+  supportsWms: boolean;
+  hasDatasets: boolean;
+  hasNestedCatalogs: boolean;
+  datasets: ThreddsDataset[];
+  catalogs: ThreddsCatalog[];
+  getAllChildDatasets: Function;
+  loadAllNestedCatalogs: Function;
+  loadNestedCatalogById: Function;
+  parentCatalog: ThreddsCatalog;
+}
+
+export interface ThreddsDataset {
+  id: string;
+  name: string;
+  url: string;
+  wmsUrl: string;
+  supportsWms: boolean;
+  isParentDataset: boolean;
+  datasets: ThreddsDataset[];
+  catalogs: ThreddsCatalog[];
+  loadAllNestedCatalogs: Function;
+}
+
+export class ThreddsStratum extends LoadableStratum(ThreddsCatalogGroupTraits) {
+  static stratumName = "thredds";
+  private threddsCatalog: ThreddsCatalog | undefined = undefined;
+
+  constructor(readonly _catalogGroup: ThreddsCatalogGroup) {
+    super();
+  }
+
+  duplicateLoadableStratum(model: BaseModel): this {
+    return new ThreddsStratum(model as ThreddsCatalogGroup) as this;
+  }
+
+  static async load(
+    catalogGroup: ThreddsCatalogGroup
+  ): Promise<ThreddsStratum | undefined> {
+    return new ThreddsStratum(catalogGroup);
+  }
+
+  @computed
+  get members(): ModelReference[] {
+    if (this.threddsCatalog === undefined) return [];
+
+    const memberIds: ModelReference[] = [];
+
+    this.threddsCatalog.catalogs.forEach((catalog: ThreddsCatalog) => {
+      memberIds.push(`${this._catalogGroup.uniqueId}/${catalog.id}`);
+    });
+
+    this.threddsCatalog.datasets.forEach((dataset: ThreddsDataset) => {
+      dataset.catalogs.forEach(c => {
+        memberIds.push(`${this._catalogGroup.uniqueId}/${c.id}`);
+      });
+      memberIds.push(`${this._catalogGroup.uniqueId}/${dataset.id}`);
+    });
+    return memberIds;
+  }
+
+  @action
+  createThreddsCatalog(catalog: ThreddsCatalog) {
+    const threddsGroup = new ThreddsCatalogGroup(
+      `${this._catalogGroup.uniqueId}/${catalog.id}`,
+      this._catalogGroup.terria
+    );
+
+    threddsGroup.setTrait(CommonStrata.definition, "name", catalog.title);
+    threddsGroup.setTrait(CommonStrata.definition, "url", catalog.url);
+    threddsGroup.terria.addModel(threddsGroup);
+  }
+
+  @action
+  async createMembers() {
+    this.threddsCatalog = await threddsCrawler(this._catalogGroup.url);
```

Comment:
We need to be able to proxy this url, but the URL parser inside the thredds-catalog-crawler doesn't like our proxy URLs:

For example:  
`proxy/_1d/http://climate-data.it.csiro.au:8080/thredds/catalog/data/authoritative/VIC-5/catalog.xml`

```bash
2urlParsingBrowser.js?a9a8:2 Uncaught (in promise) TypeError: Failed to construct 'URL': Invalid URL
    at Requestor.parseUrl (urlParsingBrowser.js?a9a8:2)
    at new Catalog (Catalog.js?08ae:19)
    at threddsCatalogCrawler (main.js?918f:4)
    at entry (entryBrowser.js?b652:13)
    at ThreddsStratum._callee$ (ThreddsCatalogGroup.ts?2f16:103)
    at tryCatch (runtime.js?96cf:63)
    at Generator.invoke [as _invoke] (runtime.js?96cf:293)
    at Generator.eval [as next] (runtime.js?96cf:118)
    at asyncGeneratorStep (ThreddsCatalogGroup.ts?2f16:18)
    at _next (ThreddsCatalogGroup.ts?2f16:18)
```

Reply:
I'm happy to take this on tomorrow :)

Reply:
fixed